pkg/controller/job/job_controller.go

```diff
@@ -1429,6 +1455,22 @@ func (jm *Controller) recordJobFinished(job *batch.Job, finishedCond *batch.JobC
 	return true
 }
 
+func (jm *Controller) recordJobSucceeded(job *batch.Job) {
+	successCriteriaMet := hasSuccessCriteriaMetCondition(job)
+	if successCriteriaMet == nil {
+		return
+	}
+	var metricReason string
+	switch successCriteriaMet.Reason {
+	case batch.JobReasonSuccessPolicy:
+		metricReason = "JobSuccessPolicy"
+	case batch.JobReasonCompletionsReached:
+		metricReason = "Completions"
+	}
+	jm.recorder.Event(job, v1.EventTypeNormal, "SuccessCriteriaMet", "Job met success criteria")
+	metrics.JobSucceededTotal.WithLabelValues(metricReason).Inc()
```

Comment:
_Rather than introducing the new metric I would prefer to just extend the existing metric JobFinishedNum for "adjust" [this line](https://github.com/kubernetes/kubernetes/blob/2ad04a0505ef13915e11d8b2b3445e514065a9bb/pkg/controller/job/job_controller.go#L1424) to take the reason from finishedCond.reason.
I know the new metric was proposed and accepted in KEP, but I think it is completely redundant with JobFinishedNum. If we are ok to add the reason at the condition level, then I think we should also be fine to add it at the metric level._

_I think otherwise the API will feel awkward - that for Failed jobs the metric reason is consistent with condition, and for succeeded job it is "", but we have a dedicated metric to split by reasons._

originally posted by @mimowo https://github.com/kubernetes/kubernetes/pull/126064#discussion_r1677379791

Reply:
> Rather than introducing the new metric I would prefer to just extend the existing metric JobFinishedNum for "adjust" [this line](https://github.com/kubernetes/kubernetes/blob/2ad04a0505ef13915e11d8b2b3445e514065a9bb/pkg/controller/job/job_controller.go#L1424) to take the reason from finishedCond.reason.
I know the new metric was proposed and accepted in KEP, but I think it is completely redundant with JobFinishedNum. If we are ok to add the reason at the condition level, then I think we should also be fine to add it at the metric level.

I understand the reason why we should extend the existing the "JobFinishedNum" metric. Actually, from the consistency perspective, I prefer to extend the `JobFinishedNum` metric instead of new `JobSucceededTotal`.

But, the existing "JobFinishedNum" metric is stable level one, and then the empty ("") reason label is reserved for the "Complete" condition. 

https://github.com/kubernetes/kubernetes/blob/f5dbc1ac156635dc7c1ab16024a1ae139993f639/pkg/controller/job/metrics/metrics.go#L58-L63

So, I'm suspecting that introducing new `JobSuccessPolicy` and `Completions` reason will break existing metric.
 Let me confirm what production readiness reviewer thinks.

@wojtek-t @soltysh Do you think this metric reason change is breaking change?
We think if we should introduce the success reasons to the exsting `jobs_finished_total` metric for the consistency between conditions and metric instead of introducing the new metric, `job_succeeded_total` as https://github.com/kubernetes/enhancements/tree/master/keps/sig-apps/3998-job-success-completion-policy#how-can-an-operator-determine-if-the-feature-is-in-use-by-workloads

```diff
-	// JobFinishedNum tracks the number of Jobs that finish. Empty reason label
-	// is used to count successful jobs.
+	// JobFinishedNum tracks the number of Jobs that finish.
	// Possible label values:
	//   completion_mode: Indexed, NonIndexed
	//   result:          failed, succeeded
-	//   reason:          "BackoffLimitExceeded", "DeadlineExceeded", "PodFailurePolicy", ""
+	//   reason:          "BackoffLimitExceeded", "DeadlineExceeded", "PodFailurePolicy", "SuccessPolicy", "Completions"
```

Reply:
I look at this way - we modify the set of reasons for the metric each time we add a new reason for the Failed or Complete condition (so its not like the set of reasons is immutable in the first place). 

If we agree that we replace the "" reason, for the Complete condition, with "CompletionsReached" (or Completions) , then I think it is natural that we update the metric accordingly.

Reply:
Uhm, that makes sense. If production readiness reviewers agree with this change, I will replace those metric reasons.

Reply:
Side Note: Through some investigations with @mimowo, we found the outdated metrics reasons comment.
The outdated comment will be fixed in https://github.com/kubernetes/kubernetes/pull/126121.

Reply:
Playing devil's advocate, what is the chance that users started relying on an empty label for succeeded jobs? I personally doubt it, so it seems like a natural extension to the metric, where we add additional information around the success reason. It seems safe, but let's hear from Wojtek as well.

Reply:
Actually this would be up to whatever our "stable" metrics policy is - @logicalhan, what are the policies regarding stable metrics and "reasons", and the semantics of those? AFAIK, we are not nearly so strict on metrics changes as we would be on API resources, but it is still technically "API".

Reply:
Is this only expanding the group of label values? If so, that's actually even compliant with stable metrics. What's not permitted with stable metrics is mutation of the data structure itself (i.e. addition or removal of labels)

Reply:
That is correct.

Reply:
Thank you for checking this discussion!

As PRR reviewers comment, extending metrics reason doesn't mean breaking changes. So, let me extend the existing `job_finished_total` metric instead of introducing new `job_succeeded_total` metric as I proposed here:

https://github.com/kubernetes/kubernetes/pull/126075#discussion_r1678442484